python/strings_udf/cpp/include/cudf/strings/udf/udf_string.hpp

```diff
@@ -0,0 +1,532 @@
+   */
+  __device__ udf_string(char const* data);
+
+  /**
+   * @brief Create a string object from a cudf::string_view
+   *
+   * The input string data is copied into the instance returned.
+   *
+   * @param str String to copy
+   */
+  __device__ udf_string(cudf::string_view const str);
+
+  /**
+   * @brief Create a string object with `count` copies of character `chr`
+   *
+   * @param count Number of times to copy `chr`
+   * @param chr Character from which to create the string
+   */
+  __device__ udf_string(cudf::size_type count, cudf::char_utf8 chr);
+
+  /**
+   * @brief Create a string object from another instance
+   *
+   * The string data is copied from the `src` into the instance returned.
+   *
+   * @param src String to copy
+   */
+  __device__ udf_string(udf_string const& src);
+
+  /**
+   * @brief Create a string object from a move reference
+   *
+   * The string data is moved from `src` into the instance returned.
+   * The `src` will have no content.
+   *
+   * @param src String to copy
+   */
+  __device__ udf_string(udf_string&& src);
+
+  __device__ ~udf_string();
+
+  __device__ udf_string& operator=(udf_string const&);
+  __device__ udf_string& operator=(udf_string&&);
+  __device__ udf_string& operator=(cudf::string_view const);
+  __device__ udf_string& operator=(char const*);
+
+  /**
+   * @brief Return the number of bytes in this string
+   */
+  __device__ cudf::size_type size_bytes() const;
+
+  /**
+   * @brief Return the number of characters in this string
+   */
+  __device__ cudf::size_type length() const;
+
+  /**
+   * @brief Return the maximum number of bytes a udf_string can hold
+   */
+  __device__ cudf::size_type max_size() const;
+
+  /**
+   * @brief Return the internal pointer to the character array for this object
+   */
+  __device__ char* data();
+  __device__ char const* data() const;
+
+  /**
+   * @brief Returns true if there are no characters in this string
+   */
+  __device__ bool is_empty() const;
+
+  /**
+   * @brief Returns an iterator that can be used to navigate through
+   *        the UTF-8 characters in this string
+   *
+   * This returns a `cudf::string_view::const_iterator` which is read-only.
+   */
+  __device__ cudf::string_view::const_iterator begin() const;
+  __device__ cudf::string_view::const_iterator end() const;
+
+  /**
+   * @brief Returns the character at the specified position
+   *
+   * This will return 0 if `pos >= length()`.
+   *
+   * @param pos Index position of character to return
+   * @return Character at position `pos`
+   */
+  __device__ cudf::char_utf8 at(cudf::size_type pos) const;
+
+  /**
+   * @brief Returns the character at the specified index
+   *
+   * This will return 0 if `pos >= length()`.
+   * Note this is read-only. Use replace() to modify a character.
+   *
+   * @param pos Index position of character to return
+   * @return Character at position `pos`
+   */
+  __device__ cudf::char_utf8 operator[](cudf::size_type pos) const;
```

Comment:
Why do we need both the operator and `at`? It looks like one is strictly a pass through to the other, no extra error-checking or anything (which is what I might have expected since the methods of this class seem to be modeled after something like `std::vector`).

Should these return pointers into the array rather than returning the value? I see in the doc that it's read-only, wondering whether there are benefits to changing that behavior.

Reply:
The `std::string` and `std::vector` have both `[]` and `at()`. The intention was that `at()` do extra checking on the index parameter and would throw exception. Perhaps that would be useful in the future but I do not want to throw exceptions in device code at this point.

Reply:
> Should these return pointers into the array rather than returning the value? I see in the doc that it's read-only, wondering whether there are benefits to changing that behavior.

It would not be good to make this writable because the argument character (UTF-8) would have to be the same number of bytes as the target otherwise the entire string array can become corrupted.

Reply:
I sort of figured we would end up wanting to avoid the exceptions route, which is why I wasn't sure if there was any intention of ever having `[]` and `at` behave differently.